warp10/src/main/java/io/warp10/continuum/gts/INTERPOLATE.java

```diff
@@ -27,23 +25,514 @@
+
+    //
+    // If gts is not bucketized and occurrences parameter is not defined, do nothing
+    //
+
+    if (!GTSHelper.isBucketized(filled) && null == params.get(PARAM_OCCURRENCES)) {
+      return filled;
+    }
+
+    //
+    // Sort GTS
+    //
+
+    GTSHelper.sort(filled);
+    //filled = GTSHelper.dedup(filled);
+
+    //
+    // If there is less than two values, we cannot interpolate, return the filled GTS now
+    //
+
+    if (filled.values < 2) {
+      return filled;
+    }
+
+    //
+    // Extract initial number of values
+    //
+
+    int nvalues = filled.values;
+
+    //
+    // Compute interpolator
+    //
+
+    double xval[] = new double[nvalues];
+    double fval[] = new double[nvalues];
+    for (int i = 0; i < nvalues; i++) {
+      xval[i] = ((Number) GTSHelper.tickAtIndex(filled, i)).doubleValue();
+      fval[i] = ((Number) GTSHelper.valueAtIndex(filled, i)).doubleValue();
+    }
+
+    PolynomialSplineFunction function = null;
+    if (null == params.get(PARAM_INTERPOLATOR)) {
+      function = (new LinearInterpolator()).interpolate(xval, fval);
+    } else {
+      switch (Interpolator.valueOf((String) params.get(PARAM_INTERPOLATOR))) {
+        case spline:
+          if (nvalues > 2) {
+            function = (new SplineInterpolator().interpolate(xval, fval));
+            break;
+          }
+        case akima:
+          if (nvalues > 4 ) {
+            function = (new AkimaSplineInterpolator().interpolate(xval, fval));
+            break;
+          }
+        case linear:
+          if (nvalues > 1) {
+            function = (new LinearInterpolator()).interpolate(xval, fval);
+            break;
+          }
+        case noop:
+          function = null;
+          break;
+      }
+    }
+
+    //
+    // Fill the result
+    // Count ticks with elevation and location
+    //
+
+    int nElevations = 0;
+    int nLocations = 0;
+
+    if (null == params.get(PARAM_OCCURRENCES)) {
+      // in this case, the result is necessarily bucketized
+      // we fill empty buckets
+
+      //
+      // Compute oldest bucket
+      //
+
+      long bucket = filled.lastbucket - filled.bucketcount * filled.bucketspan;
+
+      if ((null != filled.elevations) && (GeoTimeSerie.NO_ELEVATION != filled.elevations[0])) {
+        nElevations++;
+      }
+
+      if ((null != filled.locations) && (GeoTimeSerie.NO_LOCATION != filled.locations[0])) {
+        nLocations++;
+      }
+
+      for (int i = 1; i < nvalues; i++) {
+        if ((null != filled.elevations) && (GeoTimeSerie.NO_ELEVATION != filled.elevations[i])) {
+          nElevations++;
+        }
+        if ((null != filled.locations) && (GeoTimeSerie.NO_LOCATION != filled.locations[i])) {
+          nLocations++;
+        }
+
+        //
+        // Move bucket passed the last tick encountered
+        //
+        while(bucket < filled.lastbucket && bucket <= filled.ticks[i-1]) {
+          bucket += filled.bucketspan;
+        }
```

Comment:
It cannot be the same as INTERPOLATE:

```
//# @endpoint https://sandbox.senx.io/api/v0/exec
NEWGTS
[ 0   -50.8  -122.0  42   100 ] ADDVALUE
[ 8   -50.2  -121.0  42   50 ] ADDVALUE
[ 51   -50.5  -120.5  42   50 ] ADDVALUE
[ 60   -50.9  -119.0  42   80 ] ADDVALUE
[ 80   -50.2  -120.0  42   0 ] ADDVALUE
[ 80   -50.2  -120.0  42   -10 ] ADDVALUE
[ 100   -49.5  -118.5  42   0 ] ADDVALUE
'inputData' RENAME 
'inputGts' STORE 

[ $inputGts bucketizer.last 200 10 30 ] BUCKETIZE
{
  'interpolator' 'akima'
  'interpolator.geo' 'spline'
  'invalid.tick.value' -Infinity
} INTERPOLATE  

```

Here you expect to fill with MINLONG all the empty buckets before and after the first and last tick.